soccer_vision_msgs/msg/Goalpost.msg

```diff
@@ -0,0 +1,42 @@
+# Describes a detected goalpost from vision modules, in image coordinates.
+
+uint8 SIDE_UNKNOWN=0
+uint8 SIDE_LEFT=1
+uint8 SIDE_RIGHT=2
+
+uint8 TEAM_UNKNOWN=0
+uint8 TEAM_OWN=1
+uint8 TEAM_OPPONENT=2
+
+# A bounding box that surrounds the goalpost in the image,
+# this does not have to include the whole goalpost, but only the 
+# part of the goalpost that was detected.
+vision_msgs/BoundingBox2D bb
+
+# Top and bottom points defining the significant axis of the post.
+# The z-axis should be ignored.
+# The points should be located at the centre along the width of the post.
+geometry_msgs/Point top
+geometry_msgs/Point bottom
+
+# Orthogonal to significant vector (in pixel)
+float32 width
+
+# Whether the post is a left or right post, when looking INTO the goal.
+# Value can be SIDE_UNKNOWN, SIDE_LEFT or SIDE_RIGHT
+uint8 side
+
+# Which team's goal, the goalpost is part of. A team's goal is the one
+# they have to defend. This can be useful if the two goals have different colors,
+# and you know whether you're looking at your own or opponent's goal.
+# Value can be TEAM_UNKNOWN, TEAM_OWN or TEAM_OPPONENT
+uint8 team
+
+# Whether the goalpost has a confidence value associated.
+bool has_confidence
```

Comment:
As we discussed on Discord idk if this is the best way to define an parameter as optional.

Reply:
Copying conversation from discord across to keep it public:

ijnek:
> Thought about using -1 (or any negative value) too, but I think when writing a subscription, it's a lot more readable to write
> 
> void topic_callback(soccer_vision_msgs::msg::Ball::SharedPtr ballMsg)
> {
>   if (ballMsg->has_confidence)
>   {
>     RCLCPP_DEBUG(get_logger(), "Confidence is %f", ballMsg->confidence);
>   }
> }
> 
> 
> than to write
> 
> void topic_callback(soccer_vision_msgs::msg::Ball::SharedPtr ballMsg)
> {
>   if (ballMsg->confidence >= 0)  // If we have a confidence value
>   {
>     RCLCPP_DEBUG(get_logger(), "Confidence is %f", ballMsg->confidence);
>   }
> }
> 
> 
> On the other hand, when it comes to communication between robots over the network, it probably makes sense to use "-1" to reduce the size of the packets, but that'll be a different msg type. 

Flova:
> I get that this is more readable, but we would need to be consistent and include it to all optional params and I haven't seen the pattern in other standard message definitions. In case of e.g. the significant axis of a goal post what would be the default there if a team just detects the bounding box? A vertical line or none?

Reply:
You're right in the sense that a bool is often not used in the ros standard messages to convey optional data. An alternative I can think of, is remove the `has_confidence` field, and using **0** to indicate an unknown confidence.

I think it's safe to assume that no one will be sending objects across that actually have confidence zero. 
Also, for teams that don't set confidence values, the default will be unknown (zero). 

float32 CONFIDENCE_UNKNOWN=0 can be a constant at the top of the msg, the following equality comparison can be used (even though its a float!)
```
 if (msg->confidence == soccer_vision_msgs::msg::goalpost::CONFIDENCE_UNKNOWN)
```

I have tested the equality comparison by publishing different confidences, and the subscription can detect the unknown confidence correctly.
```
[INFO] [1632532099.973111772] [use_goalpost]: Received confidence: 0.000000
[INFO] [1632532099.973524778] [use_goalpost]: Confidence is UNKNOWN: true
```
```
[INFO] [1632532138.732930328] [use_goalpost]: Received confidence: 0.010000
[INFO] [1632532138.733119305] [use_goalpost]: Confidence is UNKNOWN: false
```

Reply:
One benefit of a *0* for the unknown default would be that it does not need any interaction in case somebody doesn't care about confidence. This can, on the other hand, lead to unexpected behavior where some module applies a confidence threshold and it drops the detections that have no confidence due to the detection method or something similar. If we for example assume a binary classification where a detector detects an object a more reasonable default would be *1* because we are certain that this is an object so we publish it, otherwise it wouldn't be published, adding the confidence of *0* to these detections would seem a bit counter-intuitive. A default confidence of *-1* could also lead to said issues with thresholding that doesn't explicitly handles this case.

But I really like the definition of a default value `soccer_vision_msgs::msg::goalpost::CONFIDENCE_UNKNOWN` like you suggested.

Reply:
> I think it's safe to assume that no one will be sending objects across that actually have confidence zero.
Also, for teams that don't set confidence values, the default will be unknown (zero).

I think a default value `0` for the confidence is problematic, especially considering "dropin" games.
There, multiple teams' robots with different code bases play together as one team. They can share their own detections using the team communication. The `ATeam` might ignore the confidence field, therefore it will always be `0`, even for excellent detections. But `TeamB` uses the confidence value regularly and drops all detections with low confidence (perhaps all balls < 0.5). So `TeamB` would ignore/drop all ball detections from `ATeam`.

Reply:
Also, I don't quite understand the example of yours.

> [INFO] [1632532099.973111772] [use_goalpost]: Received confidence: 0.000000
[INFO] [1632532099.973524778] [use_goalpost]: Confidence is UNKNOWN: true

I thought, using this method, it would be possible to distinguish between the `0.0` set by default and a `0.0` explicitly set? In that case, the example does not suffice.

Therefore, I prefer the `has_confidence` or `has_X` bool for optional values

Reply:
I think the idea was that the exact __0__ float is not likely to occur when you use a confidence estimation. It would most likely be something like 0.0000132 . But I find it never the less risky to take a value from the definition space to store this information, except the __1__ because it would have no impact, because if you don't use a model that estimates the confidence you fully need to trust the model and if you apply an algorithm that expects a confidence value it may be not optimal, but doesn't break things.

Reply:
In case, the default value is `1`: In my team-comm example from above, this would result in `TeamB` almost always trusting the ball of `ATeam` more than the own.
But I would guess, it would non the less be the task of `TeamB` to "responibly" rate the balls somehow, as a team-comm ball will probably always have more error (due to self-localizations of both robots) than the own detection and would only be used as a "estimate" in case `TeamB` has not seen the ball for a while.

Reply:
So, I didn't know this, but turns out you can set [default values for msg fields in ROS2](Ihttps://docs.ros.org/en/foxy/Concepts/About-ROS-Interfaces.html#field-default-value)

This means that we can have

```
# Constant
CONFIDENCE_UKNOWN=-1

# Field
float32 confidence -1
```

I think this is the best solution.

Reply:
I've tested that an equality comparison works fine for -1, too.

Reply:
I'll update this if we agree.

Reply:
Now, I got the idea of using [NaN](https://en.wikipedia.org/wiki/NaN?wprov=sfla1) instead? But I'm also totally fine with -1

Reply:
It seems like [NaN is not supported ](https://github.com/ros2/rosidl/issues/351), and i've never seen it being used for msgs. I actually just tried out using NaN, nan, .NAN, but all of them fail on the build.

Reply:
I've made the change on this branch now, @Flova could you please click on "resolve conversation", if you reckon this is resolved?